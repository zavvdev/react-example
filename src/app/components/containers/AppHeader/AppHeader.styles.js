import { createUseStyles } from "react-jss";

export const useAppHeaderStyles = createUseStyles((theme) => ({
  root: {
    padding: "20px 0",
    display: "flex",
    justifyContent: "space-between",
    alignItems: "center",
    borderBottom: `1px solid ${theme.borderColor}`,
    "@media screen and (max-width: 500px)": {
      flexDirection: "column",
    },
  },
  logo: {
    textTransform: "uppercase",
    fontWeight: "lighter",
    color: "inherit",
    textDecoration: "none",
    fontSize: "24px",
    "@media screen and (max-width: 500px)": {
      margin: "16px 0",
    },
  },
  side: {
    flex: 1,
    display: "flex",
    alignItems: "center",
    "&:last-child": {
      justifyContent: "flex-end",
    },
  },
  cart: {
    color: "inherit",
  },
}));
