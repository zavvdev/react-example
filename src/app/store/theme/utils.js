import { localStorageService } from "app/services/LocalStorageService";
import { IS_DARK_THEME_LOCAL_STORAGE_KEY } from "app/store/theme/config";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(IS_DARK_THEME_LOCAL_STORAGE_KEY, `${darkModeState}`);
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    IS_DARK_THEME_LOCAL_STORAGE_KEY,
  );
  return !!JSON.parse(rawStorageValue);
};
