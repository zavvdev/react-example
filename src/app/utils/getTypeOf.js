export function getTypeOf(object) {
  const objectType = typeof object;
  let returnType = objectType;
  if (objectType === "object") {
    const stringifiedObject = Object.prototype.toString.call(object);
    returnType = stringifiedObject
      .replace(/^\[object (\S+)\]$/, "$1")
      .toLowerCase();
  }
  return returnType;
}
