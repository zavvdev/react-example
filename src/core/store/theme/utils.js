import { LOCAL_STORAGE_KEYS } from "core/config/localStorage";
import { localStorageService } from "core/services/LocalStorageService";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
    `${darkModeState}`,
  );
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
  );
  return !!JSON.parse(rawStorageValue);
};
