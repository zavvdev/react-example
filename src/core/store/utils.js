import { all, call, spawn } from "redux-saga/effects";
import { errorTrackingService } from "core/services/ErrorTrackingService";

export const combineMiddleware = (middlewareList) => function* combiner() {
  const getRootGenerator = (middleware) => function* rootGenerator() {
    while (true) {
      try {
        yield call(middleware);
        break;
      } catch (e) {
        errorTrackingService.reportError(e);
      }
    }
  };

  yield all(middlewareList.map((middleware) => {
    return spawn(getRootGenerator(middleware));
  }));
};
