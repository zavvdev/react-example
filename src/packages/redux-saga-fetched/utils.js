import { ACTION_TYPES, DATA_STATUS_TYPES } from "packages/redux-saga-fetched/config";

export const createActionType = ({ createdKey, actionTypePattern }) => {
  return `${createdKey}_${actionTypePattern}`;
};

export const createKey = (key) => key.join("_");

export const createActionTypePatterns = (domain) => ({
  request: `${domain}@${ACTION_TYPES.request}`,
  success: `${domain}@${ACTION_TYPES.success}`,
  failure: `${domain}@${ACTION_TYPES.failure}`,
  invalidate: `${domain}@${ACTION_TYPES.invalidate}`,
  reset: `${domain}@${ACTION_TYPES.reset}`,
});

export const createState = ({
  isLoading,
  isFetching,
  isLoaded,
  isError,
  isValid,
  status,
  data,
}) => ({
  isLoading,
  isFetching,
  isLoaded,
  isError,
  isValid,
  status,
  data,
});

export const createResetState = () => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: false,
  isError: false,
  isValid: false,
  status: DATA_STATUS_TYPES.reset,
  data: null,
});

export const createRequestState = ({ state, payload }) => {
  const isLoading = !state[payload.createdKey]?.data;
  const isFetching = !!state[payload.createdKey]?.data;
  return createState({
    isLoading,
    isFetching,
    isLoaded: false,
    isError: false,
    isValid: false,
    status: isLoading ? DATA_STATUS_TYPES.loading : DATA_STATUS_TYPES.fetching,
    data: state[payload.createdKey]?.data || null,
  });
};

export const createSuccessState = ({ payload }) => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: true,
  isError: false,
  isValid: true,
  status: DATA_STATUS_TYPES.loaded,
  data: payload.data,
});

export const createFailureState = ({ state, payload }) => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: false,
  isError: true,
  isValid: false,
  status: DATA_STATUS_TYPES.error,
  data: state[payload.createdKey]?.data || null,
});

export const createInvalidateState = ({ state, payload }) => createState({
  isLoading: state[payload.createdKey]?.isLoading,
  isFetching: state[payload.createdKey]?.isFetching,
  isLoaded: state[payload.createdKey]?.isLoaded,
  isError: state[payload.createdKey]?.isError,
  isValid: false,
  status: state[payload.createdKey]?.status,
  data: state[payload.createdKey]?.data,
});
