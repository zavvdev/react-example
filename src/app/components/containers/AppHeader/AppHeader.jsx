import { useTranslation } from "react-i18next";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { useAppHeaderStyles } from "app/components/containers/AppHeader/AppHeader.styles";
import { ThemeSwitch } from "app/components/shared/ThemeSwitch/ThemeSwitch";
import { Typography } from "app/components/shared/Typography/Typography";
import { I18N_NAMESPACES } from "app/i18n/config";
import { GENERAL_ROUTES } from "app/router/config";
import { selectIsDarkMode } from "app/store/theme/selectors";
import { toggleDarkMode } from "app/store/theme/slice";
import { selectCartBooksLength } from "cart/gateway/output";

export function AppHeader() {
  const dispatch = useDispatch();
  const classes = useAppHeaderStyles();
  const isDarkMode = useSelector(selectIsDarkMode);
  const { t } = useTranslation(I18N_NAMESPACES.common);
  const cartBooksLength = useSelector(selectCartBooksLength);

  const handleChangeTheme = () => {
    dispatch(toggleDarkMode());
  };

  return (
    <div className={classes.root}>
      <div className={classes.side}>
        <ThemeSwitch isDark={isDarkMode} onToggle={handleChangeTheme} />
      </div>
      <Link className={classes.logo} to={GENERAL_ROUTES.dashboard}>
        <Typography tag="span">{t("appName")}</Typography>
      </Link>
      <div className={classes.side}>
        <Typography tag="div">
          <Link className={classes.cart} to={GENERAL_ROUTES.cart}>
            {t("appHeader.cart", {
              count: cartBooksLength,
            })}
          </Link>
        </Typography>
      </div>
    </div>
  );
}
