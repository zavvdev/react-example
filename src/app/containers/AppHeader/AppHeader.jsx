import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { useAppHeaderStyles } from "app/containers/AppHeader/AppHeader.styles";
import { GENERAL_ROUTES } from "app/router/config";
import { useAppTranslation } from "app/i18n/useAppTranslation";
import { cartSelectors } from "cart/gateway/output";

export function AppHeader() {
  const { t } = useAppTranslation();
  const classes = useAppHeaderStyles();
  const cartBooksLength = useSelector(cartSelectors.selectCartBooksLength);

  return (
    <div className={classes.root}>
      <Link className={classes.logo} to={GENERAL_ROUTES.books}>
        <span>{t("appName")}</span>
      </Link>
      <Link to={GENERAL_ROUTES.cart}>
        {t("appHeader.cart", {
          count: cartBooksLength,
        })}
      </Link>
    </div>
  );
}
