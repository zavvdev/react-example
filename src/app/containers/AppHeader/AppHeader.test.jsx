import { AppHeader } from "app/containers/AppHeader/AppHeader";
import { testUtils } from "__tests__/utils";

test("should render AppHeader", () => {
  const result = testUtils.render(<AppHeader />);
  expect(result.asFragment()).toMatchSnapshot();
});
