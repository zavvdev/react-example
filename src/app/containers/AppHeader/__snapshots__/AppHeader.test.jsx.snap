// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader container should render AppHeader 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="side"
    >
      <button
        class="root"
        type="button"
      >
        <div
          class="switcher"
        >
          themeSwitch.light
        </div>
      </button>
    </div>
    <a
      class="logo"
      href="/books"
    >
      <span
        class="root"
      >
        appName
      </span>
    </a>
    <div
      class="side"
    >
      <div
        class="root"
      >
        <a
          class="cart"
          href="/cart"
        >
          appHeader.cart
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
